### ISSUE TITLE
[bug] EMFILE: "Too many open files" when loading org-roam

### ISSUE THREAD

**Issue by troiganto** [chipsenkbeil/org-roam.nvim#62]
I use org-roam.nvim in conjunction with lazy.nvim as a package manager. Because org-roam.nvim is a rather slow plugin to load, it gets lazy-loaded whenever i use one of its keymaps and no earlier.

Unfortunately, there seems to be some sort of concurrency issue with it. Whenever I start vim and hit the `find_node` keymap, I get the following error:

```
E5108: Error executing lua: vim/loader.lua:0: cannot open ~/.local/share/nvim/lazy/org-roam/lua/org-roam/api.lua: Too many open files
stack traceback:
        [C]: in function 'error'
        vim/loader.lua: in function <vim/loader.lua:0>
        [C]: in function 'require'
        ~/.local/share/nvim/lazy/org-roam/lua/org-roam.lua:55: in function <~/.local/share/nvim/lazy/org-roam/lua/org-roam.lua:54>
        ~/.local/share/nvim/lazy/org-roam/lua/org-roam.lua:36: in function '__index'
        ~/.local/share/nvim/lazy/org-roam/lua/org-roam/setup/keybindings.lua:177: in function <~/.local/share/nvim/lazy/org-roam/lua/org-roam/setup/keybindings.lua:169>
EMFILE: too many open files: ~/Documents/org/daily/2024-08-14.org
```

- The error does not occur if I call `:Lazy load org-roam` and then use the keymap.
- The error does not occur if I manually open an org file (loads both orgmode and org-roam) and then use the `find_node` keymap.
- The error _does_ occur if I open the orgmode agenda (loads only orgmode) and then use `find_node`.
- It also occurs if I use any other keymap, e.g. "go to dailies directory" or "open capture prompt".

I haven't had the opportunity to take a deep look into the issue, but based on the reported line numbers, I assume the issue is the forced directory sweep that org-roam.nvim does at startup: https://github.com/chipsenkbeil/org-roam.nvim/blob/01b4666df85a38d4dbb2ea689bef0390963ffa6e/lua/org-roam/setup/database.lua#L21-L33

**Comment by chipsenkbeil** [comment]
How many files do you have? I tested this with a somewhat sizeable roam repo I found on Github to see load times, and didn't hit this issue, so surprised to hear about it. The roam plugin tries to be lazy and only load the database and scan for files on first usage, which is why any of the bindings are trigger it for you.

And you don't have any symlinks or anything within the roam directory?

Based on your pathing, are you on Mac or Linux?

**Comment by troiganto** [comment]
Hi @chipsenkbeil, thanks for the quick response! Indeed I am on Linux. Fedora 40, in case it matters.

The directory contains no symbolic links — `find -P -type l` shows no results in my org directory.

The number of files is ~1k, only slightly depending on whether you ignore non-org files or not.
```sh
$ find . -type f | wc -l
1108
$ find . -name '*.org' | wc -l
1014
```

The fact that the plugin tries to be lazy makes sense and is IMO a good choice. The fact that it hits a "too many open files" error is still somewhat surprising to me. It suggests that either many files are processed in parallel, or that files are not closed after processing.

**Comment by chipsenkbeil** [comment]
@troiganto we make use of the orgmode plugin's `OrgFiles` instance, which I believe loads everything into memory, but I don't think it maintains the file descriptors.

I didn't test the large org repo on Linux - only Mac - so maybe I would have hit that issue earlier. We do try to load files in parallel to reduce processing times. Would need to be able to reproduce the issue first, and then could dig into why and how to work around it.

Is your org collection public? I'm assuming not, but wanted to check in case we could use it here to test and find how to solve this.

**Comment by troiganto** [comment]
Hi, sorry for the delayed response, things have been busy. Unfortunately, the org collection is indeed private. But the reason for the large amount of files is not that there's a lot of content. I've been following an idea [shown by the author of emacs org-roam](https://blog.jethro.dev/posts/how_to_take_smart_notes_org/), where he "abuses" pages as a sort of pseudo-tags whose backlinks show all pages tagged with it.

So a lot of my org files are actually almost empty and only serve as a link target. I'm sure it wouldn't be too hard to create a test setup locally with some scripting. Obviously not for unit testing, but at least to resolve this issue.

I've read a bit more of the code and think I have a better handle on what happens now. The function I've cited above calls [`Database:load()`](https://github.com/chipsenkbeil/org-roam.nvim/blob/01b4666df85a38d4dbb2ea689bef0390963ffa6e/lua/org-roam/database.lua#L131), which calls [`Loader:load()`](https://github.com/chipsenkbeil/org-roam.nvim/blob/01b4666df85a38d4dbb2ea689bef0390963ffa6e/lua/org-roam/database/loader.lua#L201), which calls [`Loader:files({force=true})`](https://github.com/chipsenkbeil/org-roam.nvim/blob/01b4666df85a38d4dbb2ea689bef0390963ffa6e/lua/org-roam/database/loader.lua#L413).

Because this means that `skip=false`, this ends up calling into orgmode's [`OrgFiles:load(true)`](https://github.com/nvim-orgmode/orgmode/blob/e3500add486b17da58ce8e42a0f799161e5761c9/lua/orgmode/files/init.lua#L33).
This creates a list of promises, each promise calling [`OrgFiles:load_file()`](https://github.com/nvim-orgmode/orgmode/blob/e3500add486b17da58ce8e42a0f799161e5761c9/lua/orgmode/files/init.lua#L33). 
This in turn punts things to [`OrgFile.load()`](https://github.com/nvim-orgmode/orgmode/blob/e3500add486b17da58ce8e42a0f799161e5761c9/lua/orgmode/files/file.lua#L57), which after some hemming and hawing calls [`utils.readfile()`](https://github.com/nvim-orgmode/orgmode/blob/e3500add486b17da58ce8e42a0f799161e5761c9/lua/orgmode/utils/init.lua#L9).

If you look at that last function, it's a callback chain of `uv.fs_open()`, `uv.fs_stat()`, `uv.fs_read()` and `uv.fs_close()`. So we've got a loooong list of those promises and then `OrgFiles` does a `Promise.all()` over them. That made me think we might run into issues in how `all()` is implemented.

*However*, if I understand the `OrgPromise` code right, promises are *eager*, i.e. [the promise callback is executed from within `Promise.new()`](https://github.com/nvim-orgmode/orgmode/blob/e3500add486b17da58ce8e42a0f799161e5761c9/lua/orgmode/utils/promise.lua#L103). That means the `Promise.all()` isn't the issue — while constructing the table of promises, we already call `fs_open()` asynchronously a thousand times without waiting for the result. I think this is how we run out of file descriptors on a very low level.

Unfortunately, i'm not sure how to proceed here. It looks like the fix would be to adjust `OrgFiles.load()` to do some sort of throttling. One could limit the number of parallel open files, or maybe some more judicious `vim.schedule()` might reduce the amount of blocking code that runs without yielding. What I fear is that vanilla orgmode usually deals with only few files, so it might be hard to justify the added complexity to the maintainer.

Maybe you have an idea how best to solve this?

**Comment by troiganto** [comment]
I've just confirmed with `ulimit -n` that on my machine, the default limit on open files is 1024. With 1014 org files, it makes sense that the limit is sometimes hit on an org file and sometimes on a Lua file that the runtime wants to load. :slightly_smiling_face:

**Comment by troiganto** [comment]
Just as a quick update, since it's been a while. I've since implemented [a very hacky version](https://github.com/troiganto/orgmode/blob/wip/benchmark-throttled-orgfiles/lua/orgmode/files/init.lua#L38-L108) of `OrgFiles` that allows limiting the number of running promises while opening all files. The code's kind of a mess, but I couldn't find any races, leaks or other bugs at least.

What's taken much longer was to put together [a benchmark](https://github.com/troiganto/orgmode/blob/wip/benchmark-throttled-orgfiles/full_benchmark.py) that helps me estimate how much speedup/slowdown we actually get from `Promise.all()` vs a slowed down approach. If you have a Python venv with `pandas` installed (it was the easiest way to wrangle all the numbers), you can run `./full_benchmark.py --files-range 100 2000 100`, it'll progressively generate temp directories with 100 files, 200 files, ... 1900 files, run `org.files:load_sync()` on them while varying the throttling parameter and measure the time for the loading and only the loading. Be warned that anything more than a run over ~4 data points is going to take a while though :sweat_smile: 

There are still issues with the benchmark (no visualization yet; all files are 0 bytes, whereas even the smallest files in my collection at least has a title and a property drawer with an ID), but I let it run last night and there's some stuff in the data, that I don't understand. So more work needs to be done I'm afraid!

**Comment by chipsenkbeil** [comment]
Very impressed with the work you've put into this, @troiganto !

If you narrowed it down to OrgFiles, this probably belongs in the orgmode plugin instead of roam, with this issue being about adopting the fix once it rolls out.

Before we do any of that, @kristijanhusak do you have any additional thoughts on this one? Something to change in roam itself?

**Comment by kristijanhusak** [comment]
I recently ran into this issue and solved it (https://github.com/nvim-orgmode/orgmode/commit/1692af5a0dd363851b7846c3636040c53f9cdc79). Can you try orgmode's latest master and see if it works?

**Comment by troiganto** [comment]
Hi @kristijanhusak thanks for chiming in! I've tried your commit and can conform that it works. Your code is also much cleaner than what I could come up with; as a primary Python dev, I'm really not used to thinking in promises and callbacks. :sweat_smile: 

I'm just gonna drop the graph here that I was making literally when you sent that comment. This way, it feels less like all that profiling was for nothing. :upside_down_face: 

![orgmode-limited-loading](https://github.com/user-attachments/assets/90c5bf8b-a964-40b1-9555-800baecc837a)\
[results.tsv](https://github.com/user-attachments/files/17104319/results.txt)

Ultimately, it looks like fully serial and fully parallel loading are equally fast. I meant to investigate that curious bump in the middle, but since the problem is solved now, there isn't much point to that anymore.

Thanks again for all the help to both of you, I've learned a lot about both projects' code bases through this issue!
